### PR TITLE
Arn 337 area selection page return regions

### PR DESCRIPTION
**Pull request by adelaidaa**
Enhance `/authentication/user/USER_2` to get the regions associated with the oasys user roles.

```json
{
    "userCode": "USER_2",
    "firstName": "User",
    "lastName": "Two",
    "email": "dtest@hmps.gsi.gov.uk",
    "regions": [
        {
            "name": "Hertfordshire",
            "code": "HFS"
        }
    ],
    "enabled": true
}
```

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt [hunk 1]
```diff
@@ -44,7 +43,7 @@ data class OasysUser(
   @JoinColumn(name = "CT_AREA_EST_CODE")
   val ctAreaEst: CtAreaEst? = null,
 
-  @OneToMany(fetch = FetchType.EAGER)
+  @OneToMany
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt:46]: i can think of why we would want to eagerly load the roles. Can someone tell if there is any specific reason for it?

File: src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt [hunk 2]
```diff
@@ -77,7 +80,12 @@ class AuthenticationService(
     throw UserNotAuthorisedException("No response from OASys authentication function for user, $username")
   }
 
-  fun userCanAccessOffenderRecord(oasysUserCode: String?, offenderId: Long?, sessionId: Long?, resource: OffenderPermissionResource?): AuthorisationDto {
+  fun userCanAccessOffenderRecord(
+    oasysUserCode: String?,
+    offenderId: Long?,
+    sessionId: Long?,
+    resource: OffenderPermissionResource?
+  ): AuthorisationDto {
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt:89]: no changes there just formatting

File: src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt [hunk 3]
```diff
@@ -97,21 +105,56 @@ class AuthenticationService(
...
+              OffenderPermissionLevel.READ_ONLY,
+              OffenderPermissionResource.SENTENCE_PLAN,
+              offenderId
+            )
+            return AuthorisationDto(
+              oasysUserCode,
+              offenderId,
+              OffenderPermissionLevel.READ_ONLY,
+              OffenderPermissionResource.SENTENCE_PLAN
+            )
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt:119]: no changes there just formatting

File: src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt [hunk 4]
```diff
@@ -97,21 +105,56 @@ class AuthenticationService(
...
+              OffenderPermissionLevel.WRITE,
+              OffenderPermissionResource.SENTENCE_PLAN,
+              offenderId
+            )
+            return AuthorisationDto(
+              oasysUserCode,
+              offenderId,
+              OffenderPermissionLevel.WRITE,
+              OffenderPermissionResource.SENTENCE_PLAN
+            )
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt:133]: no changes there just formatting

File: src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt [hunk 5]
```diff
@@ -97,21 +105,56 @@ class AuthenticationService(
...
           AuthorisationStatus.AuthorisationState.NO_ACCESS -> {
             logAuthFailure(oasysUserCode, "Unauthorised", offenderId)
-            return AuthorisationDto(oasysUserCode, offenderId, OffenderPermissionLevel.UNAUTHORISED, OffenderPermissionResource.SENTENCE_PLAN)
+            return AuthorisationDto(
+              oasysUserCode,
+              offenderId,
+              OffenderPermissionLevel.UNAUTHORISED,
+              OffenderPermissionResource.SENTENCE_PLAN
+            )
           }
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt:143]: no changes there just formatting

File: src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt [hunk 6]
```diff
@@ -97,21 +105,56 @@ class AuthenticationService(
...
+              oasysUserCode,
+              offenderId,
+              authStatus.state
+            )
+            return AuthorisationDto(
+              oasysUserCode,
+              offenderId,
+              OffenderPermissionLevel.UNAUTHORISED,
+              OffenderPermissionResource.SENTENCE_PLAN
+            )
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt:157]: formatting

File: src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt [hunk 7]
```diff
@@ -121,7 +164,12 @@ class AuthenticationService(
...
     }
     logAuthFailure(oasysUserCode, "No response from OASys", offenderId)
-    return AuthorisationDto(oasysUserCode, offenderId, OffenderPermissionLevel.UNAUTHORISED, OffenderPermissionResource.SENTENCE_PLAN)
+    return AuthorisationDto(
+      oasysUserCode,
+      offenderId,
+      OffenderPermissionLevel.UNAUTHORISED,
+      OffenderPermissionResource.SENTENCE_PLAN
+    )
   }
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt:173]: formatting

File: src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt [hunk 8]
```diff
@@ -163,6 +216,13 @@ class AuthenticationService(
...
-    return UserDto.from(user)
+
+    return UserDto.from(user, user.toRegions())
+  }
+
+  fun OasysUser.toRegions(): Set<String> {
+    val ctAreaEstCode = this?.roles?.mapNotNull { it.ctAreaEstCode }?.distinct()
+    if (ctAreaEstCode?.isEmpty()!!) return emptySet()
+    return ctAreaEstCode?.let { areaRepository.findCtAreaEstByCtAreaEstCodes(it.toSet()) }!!
   }
```
Comment (adelaidaa) [inline src/main/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationService.kt:227]: new function to get the areas

File: src/test/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationServiceTest.kt [hunk 9]
```diff
@@ -17,6 +17,7 @@ import uk.gov.justice.digital.oasys.jpa.entities.AreaEstUserRole
 import uk.gov.justice.digital.oasys.jpa.entities.CtAreaEst
```
Comment (adelaidaa) [inline src/test/kotlin/uk/gov/justice/digital/oasys/services/AuthenticationServiceTest.kt:17]: forgive the formatting in this file also. sorry

File: src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt (hunk 1 above)
Comment (LucasCairns) [reply src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt:46]: Might be misunderstanding here - I suppose `FetchType.EAGER` be more performant if we were using using the roles more often than not? But it shouldn't break by changing this to `LAZY` right? we can keep an eye on our usage of this field and revert back if we need to 🤔

**Review by LucasCairns — DISMISSED** [review]
LGTM 👍

**Review by wood-n — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt (hunk 1 above)
Comment (jezhiggins) [reply src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt:46]: Lazy fetching can go wonky if you read the lazily loaded field outside of the original transaction. The fix then is to annotate the top level call with @Transactional. 
For the case we're talking about here, and more generally given the number of roles will generally be small, I think eager-loading would be fine.

Comment (SWhitlam) [reply src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt]: I think we did this to reduce the number of queries JPA was making. IIRC this was used by the old (now deprecated) oasys auth functions so it made more sense to eagerly load them as they were always required.

**Review by SWhitlam — DISMISSED** [review]

File: src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt (hunk 1 above)
Comment (adelaidaa) [reply src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt]: @SWhitlam this roles are not used any longer as we are getting the regions from the `ACTIVE_USER_AREA_EST_ROLE_VW`, should i remove the roles from here then?

Comment (SWhitlam) [reply src/main/kotlin/uk/gov/justice/digital/oasys/jpa/entities/OasysUser.kt]: Yes I think so

**Review by SWhitlam — APPROVED** [review]

**Pull request merged**